sam.sh

```diff
@@ -11,74 +11,89 @@
-        else
-            echo "SAM identities validated from ${SSH_CONNECTION%%' '*}:";
-        fi
-        ssh-add -l;
-        unset DEBUG; 
-        unset SA_Stat; 
-        unset SAM_ENV_FILE; 
-        unset SAM_AUTH_SOCK;
-        unset SAM_KEYS;
-        unset SAM_CONFIRM;
-        unset SAM_DISPLAY;
-    }
+  bye_sam () {
+    if [ "$SSH_AUTH_SOCK" = "$SAM_AUTH_SOCK" ]; then
+      echo "SAM($SSH_AGENT_PID) identities validated from localhost:";
+    else
+      echo "SAM identities validated from ${SSH_CONNECTION%%' '*}:";
+    fi
+    ssh-add -l;
+    unset DEBUG; 
+    unset SA_Stat; 
+    unset SAM_ENV_FILE; 
+    unset SAM_AUTH_SOCK;
+    unset SAM_KEYS;
+    unset SAM_CONFIRM;
+    unset SAM_DISPLAY;
+  }
 
-    #DEBUG="YES";
-    if [ -f /usr/bin/ssh-askpass ] && [ "$DISPLAY" ]; then
-        export SSH_ASKPASS=${SSH_ASKPASS-/usr/bin/ssh-askpass};
-        SAM_CONFIRM=${SSH_ASKPASS:+'-c'};
+  #From ssh-add man-page
+  # -l      Lists fingerprints of all identities currently represented by the agent.
+  # Exit status is:   0 ssh-agent is reachable and has at least one key.
+  #                   1 ssh-agent is reachable and has no keys
+  #                   2 ssh-add is unable to contact the authentication agent.
+  ssh-add -l &> /dev/null; SA_Stat=$?;
+  if [ $SA_Stat -eq 2 ]; then
+    #No agent in env, try stored agent
+    if [ -e $SAM_ENV_FILE ]; then
+      [ -n "$DEBUG" ] && echo "Using stored agent, is it viable?";
+      source $SAM_ENV_FILE;
+      ssh-add -l &> /dev/null; SA_Stat=$?;
     fi
+  else
+    [ "$DEBUG" ] && echo "agent from env is available.";
+  fi
 
-    #From ssh-add man-page
-    # -l      Lists fingerprints of all identities currently represented by the agent.
-    # Exit status is:   0 ssh-agent is reachable and has at least one key.
-    #                   1 ssh-agent is reachable and has no keys
-    #                   2 ssh-add is unable to contact the authentication agent.
-    ssh-add -l &> /dev/null; SA_Stat=$?;
-    if [ $SA_Stat -eq 2 ]; then
-        #No agent in env, try stored agent
-        if [ -e $SAM_ENV_FILE ]; then
-            [ -n "$DEBUG" ] && echo "Using stored agent, is it viable?";
-            source $SAM_ENV_FILE;
-            ssh-add -l &> /dev/null; SA_Stat=$?;
-        fi
+  [ -n "$DEBUG" ] && echo -n "SAM branching on condition: $SA_Stat, ";
+  case $SA_Stat in
+    2)  [ "$DEBUG" ] && echo "agent unreachable, start a new one.";
+      eval "$(ssh-agent -k &> /dev/null)";
+      killall ssh-agent 2>/dev/null;
+      rm -f $SAM_AUTH_SOCK
+      eval "$(ssh-agent -a $SAM_AUTH_SOCK -s | tee $SAM_ENV_FILE)";
+      ;&
+    *)  [ "$DEBUG" ] && echo "Check/add keys to agent.";
+  esac
+  SAM_KEYS=$(ssh-add -l);
+  KEYS=""
+  for key in "$@"; do
+    [ ! -f "$key" ] && echo "cannot find $key" && continue
+    [ "${SAM_KEYS#*$key}" != "$SAM_KEYS" ] && echo "$key already added" && continue
+    if [ -z "$KEYS" ]; then
```

Comment:
do you know how to do this in one command?
that is, I want KEYS=key1 key2 etc. with the right spaces.